In [1]:
# Importing necessary libraries
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder

In [2]:
# Loading the Pokemon data set
df = pd.read_csv("pokemon.csv")

# Prints the first five rows of dataset
df.head(5)

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

# About the Data set
This data set includes 721 Pokemon, including their number, name, first and second type, and basic stats: HP, Attack, Defense, Special Attack, Special Defense, and Speed.
## Columns
1. '#' : PokeDex index number
2. Name : Name of the Pokemon
3. Type 1 : Type of pokemon
4. Type 2 : Other Type of Pokemon
5. Total : Sum of Attack, Sp. Atk, Defense, Sp. Def, Speed and HP
6. HP : Hit Points
7. Attack : Attack Strength
8. Defense : Defensive Strength
9. Sp. Atk : Special Attack Strength
10. Sp. Def : Special Defensive Strength
11. Speed : Speed
12. Generation : Number of generation
13. LegendaryTrue if Legendary Pokemon False if not (more revision on mythical vs legendary needed)

In [3]:
# Prints dimension of dataset (Number of rows, Number of columns)
df.shape

(800, 13)

In [4]:
# Gives information about number of not null values and data type correspoding to a column
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
#             800 non-null int64
Name          800 non-null object
Type 1        800 non-null object
Type 2        414 non-null object
Total         800 non-null int64
HP            800 non-null int64
Attack        800 non-null int64
Defense       800 non-null int64
Sp. Atk       800 non-null int64
Sp. Def       800 non-null int64
Speed         800 non-null int64
Generation    800 non-null int64
Legendary     800 non-null bool
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


In [5]:
# Prints the number of null values in each column
df.isnull().sum()

#               0
Name            0
Type 1          0
Type 2        386
Total           0
HP              0
Attack          0
Defense         0
Sp. Atk         0
Sp. Def         0
Speed           0
Generation      0
Legendary       0
dtype: int64

On observing the above two cells we can see that there are 386 Null values are present in Type 2 column so first we have to remove these null values. Rest of the columns do not have null value.

# Data Cleaning

In [6]:
# Replaces the Null Values with the mode of Type 2 column
df["Type 2"].fillna(df["Type 2"].mode()[0], inplace = True)

In [7]:
# Crosschecks the number of null values in each column
df.isnull().sum()

#             0
Name          0
Type 1        0
Type 2        0
Total         0
HP            0
Attack        0
Defense       0
Sp. Atk       0
Sp. Def       0
Speed         0
Generation    0
Legendary     0
dtype: int64

Now we can see that all columns are free from Null Values, Cleaning part is done...!!!!!

In [8]:
# Drops the columns which are not important
df.drop(["#", "Name"], axis=1, inplace=True)

# Prints the first five rows with remaining columns
df.head(5)

Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
0  Grass  Poison    318  45      49       49       65       65     45   
1  Grass  Poison    405  60      62       63       80       80     60   
2  Grass  Poison    525  80      82       83      100      100     80   
3  Grass  Poison    625  80     100      123      122      120     80   
4   Fire  Flying    309  39      52       43       60       50     65   

   Generation  Legendary  
0           1      False  
1           1      False  
2           1      False  
3           1      False  
4           1      False

### Transforming Nominal Attributes

In [9]:
# Prints the unique values present in Type 1 Column
df["Type 1"].unique()

array(['Grass', 'Fire', 'Water', 'Bug', 'Normal', 'Poison', 'Electric',
       'Ground', 'Fairy', 'Fighting', 'Psychic', 'Rock', 'Ghost', 'Ice',
       'Dragon', 'Dark', 'Steel', 'Flying'], dtype=object)

In [10]:
# Prints the number of unique values present in Type 1 Column
len(df["Type 1"].unique())

18

There are 18 distinct categories in column Type 1. We can now generate a label encoding scheme for mapping each category to a numeric value.

In [11]:
# Create the Labelencoder object
t1_le = LabelEncoder()

# Converts the categorical columns into numeric
df["Type 1_Label"] = t1_le.fit_transform(df["Type 1"])

# Prints the first five rows after modification
df.head(5)

Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
0  Grass  Poison    318  45      49       49       65       65     45   
1  Grass  Poison    405  60      62       63       80       80     60   
2  Grass  Poison    525  80      82       83      100      100     80   
3  Grass  Poison    625  80     100      123      122      120     80   
4   Fire  Flying    309  39      52       43       60       50     65   

   Generation  Legendary  Type 1_Label  
0           1      False             9  
1           1      False             9  
2           1      False             9  
3           1      False             9  
4           1      False             6

In [12]:
# Shows the mapping of Numerical Labels to Categorical Labels
Type1_mappings = {index: label for index, label in enumerate(t1_le.classes_)}
Type1_mappings

{0: 'Bug',
 1: 'Dark',
 2: 'Dragon',
 3: 'Electric',
 4: 'Fairy',
 5: 'Fighting',
 6: 'Fire',
 7: 'Flying',
 8: 'Ghost',
 9: 'Grass',
 10: 'Ground',
 11: 'Ice',
 12: 'Normal',
 13: 'Poison',
 14: 'Psychic',
 15: 'Rock',
 16: 'Steel',
 17: 'Water'}

Similarly we have to do for column "Type 2" column

In [13]:
# Prints the unique values present in Type 2 Column
df["Type 2"].unique()

array(['Poison', 'Flying', 'Dragon', 'Ground', 'Fairy', 'Grass',
       'Fighting', 'Psychic', 'Steel', 'Ice', 'Rock', 'Dark', 'Water',
       'Electric', 'Fire', 'Ghost', 'Bug', 'Normal'], dtype=object)

In [14]:
# Prints the number of unique values present in Type 2 Column
len(df["Type 2"].unique())

18

There are 18 distinct categories in column Type 2. We can now generate a label encoding scheme for mapping each category to a numeric value.

In [15]:
# Create the Labelencoder object
t2_le = LabelEncoder()

# Converts the categorical columns into numeric
df["Type 2_Label"] = t2_le.fit_transform(df["Type 2"])

# Prints the first five rows after modification
df.head(5)

Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
0  Grass  Poison    318  45      49       49       65       65     45   
1  Grass  Poison    405  60      62       63       80       80     60   
2  Grass  Poison    525  80      82       83      100      100     80   
3  Grass  Poison    625  80     100      123      122      120     80   
4   Fire  Flying    309  39      52       43       60       50     65   

   Generation  Legendary  Type 1_Label  Type 2_Label  
0           1      False             9            13  
1           1      False             9            13  
2           1      False             9            13  
3           1      False             9            13  
4           1      False             6             7

In [16]:
# Shows the mapping of Numerical Labels to Categorical Labels
Type2_mappings = {index: label for index, label in enumerate(t2_le.classes_)}
Type2_mappings

{0: 'Bug',
 1: 'Dark',
 2: 'Dragon',
 3: 'Electric',
 4: 'Fairy',
 5: 'Fighting',
 6: 'Fire',
 7: 'Flying',
 8: 'Ghost',
 9: 'Grass',
 10: 'Ground',
 11: 'Ice',
 12: 'Normal',
 13: 'Poison',
 14: 'Psychic',
 15: 'Rock',
 16: 'Steel',
 17: 'Water'}

### Transforming Ordinal Attributes

In [17]:
# Prints the unique values present in Generation Column
df["Generation"].unique()

array([1, 2, 3, 4, 5, 6], dtype=int64)

In [18]:
# Prints the unique values present in Legendary Column
df["Legendary"].unique()

array([False,  True])

In [19]:
# Create the Labelencoder object
lg_le = LabelEncoder()

# Converts the categorical columns into numeric
df["Legendary_Label"] = lg_le.fit_transform(df["Legendary"])

# Prints the first five rows after modification
df.head(5)

Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
0  Grass  Poison    318  45      49       49       65       65     45   
1  Grass  Poison    405  60      62       63       80       80     60   
2  Grass  Poison    525  80      82       83      100      100     80   
3  Grass  Poison    625  80     100      123      122      120     80   
4   Fire  Flying    309  39      52       43       60       50     65   

   Generation  Legendary  Type 1_Label  Type 2_Label  Legendary_Label  
0           1      False             9            13                0  
1           1      False             9            13                0  
2           1      False             9            13                0  
3           1      False             9            13                0  
4           1      False             6             7                0

In [20]:
# Shows the mapping of Numerical Labels to Categorical Labels
Legendary_mappings = {index: label for index, label in enumerate(lg_le.classes_)}
Legendary_mappings

{0: False, 1: True}

Till Now, we just converted categories to numerical labels, why we need this now? The reason is quite simple. Considering column Type 1, if we directly fed the Type 1_Label attribute as a feature in a machine learning model, it would consider it to be a continuous numeric feature thinking value 10 (Ground) is greater than 6 (Fire) but that is meaningless because the Ground type is certainly not bigger or smaller than Fire, these are essentially different values or categories which cannot be compared directly. Hence we need an additional layer of encoding schemes where dummy features are created for each unique value or category out of all the distinct categories per attribute.

## Encoding

### One-hot Encoding of Nominal Attributes

In [21]:
# Encode Type 1 labels using one-hot encoding scheme

# Creating one hot encoder object
t1_ohe = OneHotEncoder()

# Returns a 2D array where number of column is number of unique labels and number of row is number of rows present
# in the dataset and it can only contain a value of 1 or 0, values as 1 (indicating it as active).
t1_feature_arr = t1_ohe.fit_transform(df[['Type 1_Label']]).toarray()

# Listing the different labels present in the column
t1_feature_labels = list(t1_le.classes_)

# Prints the labels
t1_feature_labels

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\preprocessing\_encoders.py:371: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)


['Bug',
 'Dark',
 'Dragon',
 'Electric',
 'Fairy',
 'Fighting',
 'Fire',
 'Flying',
 'Ghost',
 'Grass',
 'Ground',
 'Ice',
 'Normal',
 'Poison',
 'Psychic',
 'Rock',
 'Steel',
 'Water']

In [22]:
# Attached a string "T1" with the labels of column Type 1

t1_labels = []
for label in t1_feature_labels:
    t1_labels.append("T1_"+label)
    
# Prints the labels after modification
t1_labels

['T1_Bug',
 'T1_Dark',
 'T1_Dragon',
 'T1_Electric',
 'T1_Fairy',
 'T1_Fighting',
 'T1_Fire',
 'T1_Flying',
 'T1_Ghost',
 'T1_Grass',
 'T1_Ground',
 'T1_Ice',
 'T1_Normal',
 'T1_Poison',
 'T1_Psychic',
 'T1_Rock',
 'T1_Steel',
 'T1_Water']

In [23]:
# Makes a Data frame with 2D array and modified labels
t1_features = pd.DataFrame(t1_feature_arr, columns=t1_labels)

# Prints the first five rows of the data frame
t1_features.head(5)

T1_Bug  T1_Dark  T1_Dragon  T1_Electric  T1_Fairy  T1_Fighting  T1_Fire  \
0     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
1     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
2     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
3     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
4     0.0      0.0        0.0          0.0       0.0          0.0      1.0   

   T1_Flying  T1_Ghost  T1_Grass  T1_Ground  T1_Ice  T1_Normal  T1_Poison  \
0        0.0       0.0       1.0        0.0     0.0        0.0        0.0   
1        0.0       0.0       1.0        0.0     0.0        0.0        0.0   
2        0.0       0.0       1.0        0.0     0.0        0.0        0.0   
3        0.0       0.0       1.0        0.0     0.0        0.0        0.0   
4        0.0       0.0       0.0        0.0     0.0        0.0        0.0   

   T1_Psychic  T1_Rock  T1_Steel  T1_Water  
0         0.0      0.0       0.0       0.0  
1         0.0      0.0       0.0       0.0  
2         0.0      0.0       0.0       0.0  
3         0.0      0.0       0.0       0.0  
4         0.0      0.0       0.0       0.0

In [24]:
# Encode Type 2 labels using one-hot encoding scheme

# Creating one hot encoder object
t2_ohe = OneHotEncoder()

# Returns a 2D array where number of column is number of unique labels and number of row is number of rows present
# in the dataset and it can only contain a value of 1 or 0, values as 1 (indicating it as active).
t2_feature_arr = t2_ohe.fit_transform(df[['Type 2_Label']]).toarray()

# Listing the different labels present in the column
t2_feature_labels = list(t2_le.classes_)

# Prints the labels
t2_feature_labels

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\preprocessing\_encoders.py:371: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)


['Bug',
 'Dark',
 'Dragon',
 'Electric',
 'Fairy',
 'Fighting',
 'Fire',
 'Flying',
 'Ghost',
 'Grass',
 'Ground',
 'Ice',
 'Normal',
 'Poison',
 'Psychic',
 'Rock',
 'Steel',
 'Water']

In [25]:
# Attached a string "T2" with the labels of column Type 2

t2_labels = []
for label in t2_feature_labels:
    t2_labels.append("T2_"+label)
    
# Prints the labels after modification   
t2_labels

['T2_Bug',
 'T2_Dark',
 'T2_Dragon',
 'T2_Electric',
 'T2_Fairy',
 'T2_Fighting',
 'T2_Fire',
 'T2_Flying',
 'T2_Ghost',
 'T2_Grass',
 'T2_Ground',
 'T2_Ice',
 'T2_Normal',
 'T2_Poison',
 'T2_Psychic',
 'T2_Rock',
 'T2_Steel',
 'T2_Water']

In [26]:
# Makes a Data frame with 2D array and modified labels
t2_features = pd.DataFrame(t2_feature_arr, columns=t2_labels)

# Prints the first five rows of the data frame
t2_features.head(5)

T2_Bug  T2_Dark  T2_Dragon  T2_Electric  T2_Fairy  T2_Fighting  T2_Fire  \
0     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
1     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
2     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
3     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
4     0.0      0.0        0.0          0.0       0.0          0.0      0.0   

   T2_Flying  T2_Ghost  T2_Grass  T2_Ground  T2_Ice  T2_Normal  T2_Poison  \
0        0.0       0.0       0.0        0.0     0.0        0.0        1.0   
1        0.0       0.0       0.0        0.0     0.0        0.0        1.0   
2        0.0       0.0       0.0        0.0     0.0        0.0        1.0   
3        0.0       0.0       0.0        0.0     0.0        0.0        1.0   
4        1.0       0.0       0.0        0.0     0.0        0.0        0.0   

   T2_Psychic  T2_Rock  T2_Steel  T2_Water  
0         0.0      0.0       0.0       0.0  
1         0.0      0.0       0.0       0.0  
2         0.0      0.0       0.0       0.0  
3         0.0      0.0       0.0       0.0  
4         0.0      0.0       0.0       0.0

### One-hot Encoding of Ordinal Attributes

In [27]:
# Encode Legendary labels using one-hot encoding scheme

# Creating one hot encoder object
lg_ohe = OneHotEncoder()

# Returns a 2D array where number of column is number of unique labels and number of row is number of rows present
# in the dataset and it can only contain a value of 1 or 0, values as 1 (indicating it as active).
lg_feature_arr = lg_ohe.fit_transform(df[['Legendary_Label']]).toarray()

# Listing the different labels present in the column
lg_feature_labels = list(lg_le.classes_)

# Prints the labels
lg_feature_labels

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\preprocessing\_encoders.py:371: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)


[False, True]

In [28]:
# Attached a string "Lg" with the labels of column Legendary

lg_labels = []
for label in lg_feature_labels:
    lg_labels.append("Lg_"+str(label))
    
# Prints the labels after modification   
lg_labels

['Lg_False', 'Lg_True']

In [29]:
# Makes a Data frame with 2D array and modified labels
lg_features = pd.DataFrame(lg_feature_arr, columns=lg_labels)

# Prints the first five rows of the data frame
lg_features.head(5)

Lg_False  Lg_True
0       1.0      0.0
1       1.0      0.0
2       1.0      0.0
3       1.0      0.0
4       1.0      0.0

In [30]:
# Encode Generation labels using one-hot encoding scheme

# Creating one hot encoder object
gen_ohe = OneHotEncoder()

# Returns a 2D array where number of column is number of unique labels and number of row is number of rows present
# in the dataset and it can only contain a value of 1 or 0, values as 1 (indicating it as active).
gen_feature_arr = gen_ohe.fit_transform(df[['Generation']]).toarray()

# Listing the different labels present in the column
gen_feature_labels = list(df["Generation"].unique())

# Prints the labels
gen_feature_labels

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\preprocessing\_encoders.py:371: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)


[1, 2, 3, 4, 5, 6]

In [31]:
# Attached a string "Gen" with the labels of column Generation

gen_labels = []
for label in gen_feature_labels:
    gen_labels.append("Gen_"+str(label))
    
# Prints the labels after modification
gen_labels

['Gen_1', 'Gen_2', 'Gen_3', 'Gen_4', 'Gen_5', 'Gen_6']

In [32]:
# Makes a Data frame with 2D array and modified labels
gen_features = pd.DataFrame(gen_feature_arr, columns=gen_labels)

# Prints the first five rows of the data frame
gen_features.head(5)

Gen_1  Gen_2  Gen_3  Gen_4  Gen_5  Gen_6
0    1.0    0.0    0.0    0.0    0.0    0.0
1    1.0    0.0    0.0    0.0    0.0    0.0
2    1.0    0.0    0.0    0.0    0.0    0.0
3    1.0    0.0    0.0    0.0    0.0    0.0
4    1.0    0.0    0.0    0.0    0.0    0.0

In [33]:
# Concatenates all the new data frame and the old data frame df 
df_ohe = pd.concat([t1_features, t2_features, df, gen_features, lg_features], axis=1)

# Prints the first five rows
df_ohe.head(5)

T1_Bug  T1_Dark  T1_Dragon  T1_Electric  T1_Fairy  T1_Fighting  T1_Fire  \
0     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
1     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
2     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
3     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
4     0.0      0.0        0.0          0.0       0.0          0.0      1.0   

   T1_Flying  T1_Ghost  T1_Grass  ...  Type 2_Label  Legendary_Label  Gen_1  \
0        0.0       0.0       1.0  ...            13                0    1.0   
1        0.0       0.0       1.0  ...            13                0    1.0   
2        0.0       0.0       1.0  ...            13                0    1.0   
3        0.0       0.0       1.0  ...            13                0    1.0   
4        0.0       0.0       0.0  ...             7                0    1.0   

   Gen_2  Gen_3  Gen_4  Gen_5  Gen_6  Lg_False  Lg_True  
0    0.0    0.0    0.0    0.0    0.0       1.0      0.0  
1    0.0    0.0    0.0    0.0    0.0       1.0      0.0  
2    0.0    0.0    0.0    0.0    0.0       1.0      0.0  
3    0.0    0.0    0.0    0.0    0.0       1.0      0.0  
4    0.0    0.0    0.0    0.0    0.0       1.0      0.0  

[5 rows x 58 columns]

In [34]:
# Drops all the columns for which we did label encoding and one-hot encoding
df_ohe.drop(["Type 1", "Type 2", "Type 1_Label", "Type 2_Label", "Legendary", "Legendary_Label", "Generation"], axis=1, inplace=True)

# Prints the first five rows
df_ohe.head(5)

T1_Bug  T1_Dark  T1_Dragon  T1_Electric  T1_Fairy  T1_Fighting  T1_Fire  \
0     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
1     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
2     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
3     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
4     0.0      0.0        0.0          0.0       0.0          0.0      1.0   

   T1_Flying  T1_Ghost  T1_Grass  ...  Sp. Def  Speed  Gen_1  Gen_2  Gen_3  \
0        0.0       0.0       1.0  ...       65     45    1.0    0.0    0.0   
1        0.0       0.0       1.0  ...       80     60    1.0    0.0    0.0   
2        0.0       0.0       1.0  ...      100     80    1.0    0.0    0.0   
3        0.0       0.0       1.0  ...      120     80    1.0    0.0    0.0   
4        0.0       0.0       0.0  ...       50     65    1.0    0.0    0.0   

   Gen_4  Gen_5  Gen_6  Lg_False  Lg_True  
0    0.0    0.0    0.0       1.0      0.0  
1    0.0    0.0    0.0       1.0      0.0  
2    0.0    0.0    0.0       1.0      0.0  
3    0.0    0.0    0.0       1.0      0.0  
4    0.0    0.0    0.0       1.0      0.0  

[5 rows x 51 columns]

In [35]:
# Expands the output display to see more columns. In the above cell we are not able to see all columns
# but in the below cell we are able to see all columns
pd.set_option('display.max_columns', None)

### Final Data after applying Feature Engineering using Label Encoding and OneHotEncoding

In [36]:
# Prints first five rows
df_ohe.head(5)

T1_Bug  T1_Dark  T1_Dragon  T1_Electric  T1_Fairy  T1_Fighting  T1_Fire  \
0     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
1     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
2     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
3     0.0      0.0        0.0          0.0       0.0          0.0      0.0   
4     0.0      0.0        0.0          0.0       0.0          0.0      1.0   

   T1_Flying  T1_Ghost  T1_Grass  T1_Ground  T1_Ice  T1_Normal  T1_Poison  \
0        0.0       0.0       1.0        0.0     0.0        0.0        0.0   
1        0.0       0.0       1.0        0.0     0.0        0.0        0.0   
2        0.0       0.0       1.0        0.0     0.0        0.0        0.0   
3        0.0       0.0       1.0        0.0     0.0        0.0        0.0   
4        0.0       0.0       0.0        0.0     0.0        0.0        0.0   

   T1_Psychic  T1_Rock  T1_Steel  T1_Water  T2_Bug  T2_Dark  T2_Dragon  \
0         0.0      0.0       0.0       0.0     0.0      0.0        0.0   
1         0.0      0.0       0.0       0.0     0.0      0.0        0.0   
2         0.0      0.0       0.0       0.0     0.0      0.0        0.0   
3         0.0      0.0       0.0       0.0     0.0      0.0        0.0   
4         0.0      0.0       0.0       0.0     0.0      0.0        0.0   

   T2_Electric  T2_Fairy  T2_Fighting  T2_Fire  T2_Flying  T2_Ghost  T2_Grass  \
0          0.0       0.0          0.0      0.0        0.0       0.0       0.0   
1          0.0       0.0          0.0      0.0        0.0       0.0       0.0   
2          0.0       0.0          0.0      0.0        0.0       0.0       0.0   
3          0.0       0.0          0.0      0.0        0.0       0.0       0.0   
4          0.0       0.0          0.0      0.0        1.0       0.0       0.0   

   T2_Ground  T2_Ice  T2_Normal  T2_Poison  T2_Psychic  T2_Rock  T2_Steel  \
0        0.0     0.0        0.0        1.0         0.0      0.0       0.0   
1        0.0     0.0        0.0        1.0         0.0      0.0       0.0   
2        0.0     0.0        0.0        1.0         0.0      0.0       0.0   
3        0.0     0.0        0.0        1.0         0.0      0.0       0.0   
4        0.0     0.0        0.0        0.0         0.0      0.0       0.0   

   T2_Water  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  Gen_1  \
0       0.0    318  45      49       49       65       65     45    1.0   
1       0.0    405  60      62       63       80       80     60    1.0   
2       0.0    525  80      82       83      100      100     80    1.0   
3       0.0    625  80     100      123      122      120     80    1.0   
4       0.0    309  39      52       43       60       50     65    1.0   

   Gen_2  Gen_3  Gen_4  Gen_5  Gen_6  Lg_False  Lg_True  
0    0.0    0.0    0.0    0.0    0.0       1.0      0.0  
1    0.0    0.0    0.0    0.0    0.0       1.0      0.0  
2    0.0    0.0    0.0    0.0    0.0       1.0      0.0  
3    0.0    0.0    0.0    0.0    0.0       1.0      0.0  
4    0.0    0.0    0.0    0.0    0.0       1.0      0.0